File: my-reader-mobile/modules/my-reader-core/android/src/main/java/com/myreader/core/MyReaderCoreModule.kt
package com.myreader.core

import com.myreader.core.uniffi.CoreFfiException
import com.myreader.core.uniffi.coreContractVersion
import com.myreader.core.uniffi.invokeCoreAsync
import com.myreader.core.uniffi.invokeCoreSync
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MyReaderCoreModule : Module() {
  private fun <T> coreCall(operation: () -> T): T = try {
    operation()
  } catch (error: CoreFfiException) {
    when (error) {
      is CoreFfiException.Core ->
        throw CodedException("CORE_ERROR", error.v1, error)
      is CoreFfiException.Sync ->
        throw CodedException("SYNC_ERROR", error.v1, error)
    }
  }

  override fun definition() = ModuleDefinition {
    Name("MyReaderCore")

    Function("coreContractVersion") {
      coreContractVersion()
    }
    Function("invokeCoreSync") { requestJson: String ->
      coreCall {
        invokeCoreSync(requestJson)
      }
    }
    AsyncFunction("invokeCoreAsync") { requestJson: String ->
      coreCall {
        invokeCoreAsync(requestJson)
      }
    }
  }
}

File: my-reader-mobile/modules/my-reader-core/ios/MyReaderCoreModule.swift
import ExpoModulesCore

private func coreCall<T>(_ operation: () throws -> T) throws -> T {
  do {
    return try operation()
  } catch let CoreFfiError.Core(message) {
    throw Exception(
      name: "MyReaderCoreException",
      description: message,
      code: "CORE_ERROR"
    )
  } catch let CoreFfiError.Sync(message) {
    throw Exception(
      name: "MyReaderCoreSyncException",
      description: message,
      code: "SYNC_ERROR"
    )
  } catch {
    throw Exception(
      name: "MyReaderCoreException",
      description: error.localizedDescription,
      code: "CORE_ERROR"
    )
  }
}

public class MyReaderCoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MyReaderCore")

    Function("coreContractVersion") {
      coreContractVersion()
    }

    Function("invokeCoreSync") {
      (requestJson: String) in
      try coreCall {
        try invokeCoreSync(requestJson: requestJson)
      }
    }

    AsyncFunction("invokeCoreAsync") {
      (requestJson: String) in
      try coreCall {
        try invokeCoreAsync(requestJson: requestJson)
      }
    }
  }
}
